xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp

```diff
@@ -547,6 +556,7 @@ void CDVDInputStreamBluray::ProcessEvent() {
   case BD_EVENT_MENU:
     CLog::Log(LOGDEBUG, "CDVDInputStreamBluray - BD_EVENT_MENU %d",
         m_event.param);
+    m_player->OnDiscNavResult(nullptr, BD_EVENT_MENU);
```

Comment:
why would this cause `IsBlurayTopMenu` to always be true (see according code in VideoPlayer) while `m_menu` changes it's state? If the bluray is not in `m_menu` state it can't be in `IsBlurayTopMenu` state either.

Reply:
Because that's the exact focus of this commit and the fix we need. 

There are 2 bluray disc types HDMV and BDJ. HDMV doesn't use java and is not important for this discussion. BDJ requires Java and you can easily tell it's a BDJ if you look inside the folder structure of a disc and see a BDJO folder with *.bdjo files. This is the vast majority of discs and it's the one that causes the problems described here. 

When you play a BDJ disc, specially a 4k one (because as far as I know almost all are BDJ), the m_menu state will always be true. This might look as a problem, but it's not in reality. The disc will have a main top menu and a pop up menu that should be accessible at all time. As described in the first commit I made last year there's no way to figure out when the popup menu is displaying or not, the disc is always setting this to true and never discard or close the menu overlay.

Now, one of the things m_menu does, while informing the player it is inside a menu and should not skip and seeking should always be zero, is that it passes the user input to BD_VK structure. This means that when you press left or right you will seek with Kodi, meaning you will move yourself in the timeline of the play item if m_menu is set to false. 
If m_menu is set to true, and you press left and right nothing will happen during the main movie playback (this is expected and it's correct way, similar to an actual bluray player), but if you press ctrl+shift+m, or if you open the Kodi OSD by pressing just M, and press on top of that small house icon, you will toggle the popup menu to show up and if there's no popup menu it will go back to main top menu. 

With your code, since you are changing the IsInMenu and changing other things, when you open the popup menu and press left and right, you will move yourself in Kodi timeline. 

So, with that explained, this is why we need 2 bools to check, one, m_menu (meaning, should the user input be passed to the disc or to kodi OSD control?) two, IsBlurayTopMenu, what IsBlurayTopMenu will do is: yes, it is inside a menu, it will be inside a menu during the entire movie, but even when inside a menu, it should be able to skip chapter and seek, we don't want the disc playback to think it is not inside a menu in any given time when playing a BDJ but we want to be able to skip and seek. 

I have compiled the changes I made and posted in the forum for people to try out. 
https://forum.kodi.tv/showthread.php?tid=359995&pid=3009533#pid3009533

try this, and try your code with a BDJO pop up menu, try to control the BDJO menu.
One good test, try to move the cursor when Resume playback (yes/no) shows up, it's very common for 4k uhd bluray.

Reply:
I know what you are trying to do. My comment here was merely why you would basically set `m_IsBlurayTopMenu` to true while `m_menu` might be false? You can't be in TopMenu when m_menu is false, can you? So this particular line did not make sense to me.

As for the "resume playback" dialog. IMO the underlying issues with that has a way different cause and your "always catch user input no matter what" only makes it work by chance. The underlying issue is, that the menu is rendered too early because VideoPlayer uses a buffer and thus is rendering video delayed while libbluray is ahead of time and already renders an overlay that's not yet supposed to be there. You can clearly see this when you add a breakpoint to the `OnMenu` action of DVDInputStreamBluray. So the correct fix for this would be to fix the timings, so that the overlay of the "resume playback" menu is rendered when the bluray is actually in the according menu state. This is basically also the reason why it's taking ages for menu actions to actually execute. So once the timings are in sync, the resume playback menu IMO would also work as expected.

Also, I think we should not trap all user input for BD-J blurays by default. The "full bluray mode" or however we might call it should be an option the user can choose/enable, because especially for limited button remotes without dedicated skip buttons it's otherwise not possible to do any seeking except via OSD. So your current solution might please some users, but annoy others.

Reply:
I'm not sure if I follow but m_menu isn't false. It is set as a value != 0 and you are correct, you can't be in TopMenu with m_menu as false. 

I had no idea about the cause but I don't see any slowdown while browsing the menus. I see a minor hickup when skipping chapters as reported in the forum. 

This is something I disagree. There's no other way around this. Or you control Kodi, or you control the disk. If you look at how Jriver does this it actually has an entire mode "bluray playback" that disables a lot of stuff, because as we know, a ton of action overlap each other. 
The user looking to play a complete bluray will want a complete bluray experience, and this overlap with the "always controlling Kodi experience". You have the simplified menu experience and the main movie experience. 
When playing a full disc, you will be controlling the disk. 

If you don't trap the user input and the user wants to move left on a popup menu during playback you will seek the time in the playback. 
The only solution to this would be: if the menu is rendered on screen user input should go to disk, if not control Kodi. This would be the optimal way, but again, we have no way to know when the bluray popup menu is rendered and closes. I looked into every single value while debugging. I spoke with libbluray creator, there's no way around this. Ace once mentioned that the only method would be using something that reads the screen, much like RetroArch does for translating game with google API, but this is way above my head and seems like a nuke to kill an ant solution.

Also, as you can see m_menu when set to on, when the menu is flushed, it never is false.

So the user do have the option: main movie mode, simplified menu mode, and bluray menu mode. 
In the third choice, he still have the option to open the OSD  menu. 
I'm working on bluray menu playback on Android and I'm using the Shield, which has the most basic remote possible. You can do everything with that remote, just hold play (that is a basic button on every remote) and it will open the OSD and you can skip, seek and open the popup menu or the main menu.
While in the main menu you can control the disk. 

Unless you have a remote with 2 pairs of [up, down, left, right, enter, menu key] there's no way around this and it's the same behavior as a standalone bluray disc. 
The solution I came up on previews commit was setting m_menu = !m_menu, but this will fail for discs where the playback doesn't set m_menu properly, or if the user closes the popup menu, or if the popup menu timeout and disappear, next time you press the popup menu key, you will be locked out. 

It's a dilema I agree and I agree that you might annoy some but you will please exactly the public that this solution is aimed for, the people that want to have a full bluray player experience.

Reply:
> I'm not sure if I follow but m_menu isn't false. It is set as a value != 0 and you are correct, you can't be in TopMenu with m_menu as false.
It's a boolean, so it can be true or false here. In case it is false IsTopMenu would still be true in your case, which technically would be wrong

> I had no idea about the cause but I don't see any slowdown while browsing the menus. I see a minor hickup when skipping chapters as reported in the forum.

You never noticed that when you clicked an action in a menu that the action would not be performed immediately but that KODI sometimes waits for the background animation/clip to finish before it executes the command? That's the thing that annoyed me most. Might be a HDMV thing though - but BD-J is also not an instant response.

I see your point and the issue with not being able to reliably detect when a menu is on. But I'm also not keen on always trapping the user input. There IMO should be a way to toggle it if one prefers keeping the possibility for skip steps and what not even though he started the movie in menu mode.
But since this change likely is nothing for v19 anymore as the release is around the corner, we have some time to think about a nice and clean solution.